gdal/doc/source/development/rfc/rfc76_ogrpythondrivers.rst

```diff
@@ -0,0 +1,397 @@
+        # and only implemented if fast feature count is possible when a
+        # spatial and/or attribute filter is set on the layer.
+        def feature_count(self, force_computation):
+            if self.attribute_filter is not None or \
+            self.geometry_filter is not None:
+                return super(Layer, self).feature_count(force_computation)
+
+            return self.count
+
+        # Optional. Should generally be exclusive with feature_count()
+        def unfiltered_feature_count(self, force_computation):
+            assert self.attribute_filter is None and self.geometry_filter is None
+            return self.count
+
+        # Required
+        def reset_reading(self):
+            self.iter = 0
+
+        # Optional, unless next_raw_feature is defined.
+        # Must then honour self.attribute_filter and self.geometry_filter
+        # attribute_filter will be None or a OGR SQL WHERE clause
+        # geometry_filter will be None or [minx, miny, maxx, maxy]
+        # def next_feature(self):
+        #     return {}
+
+        # Required, unless nextFeature is defined.
+        def next_raw_feature(self):
+            if self.iter == self.count:
+                return None
+            self.iter += 1
+
+            # A GeoJSON Feature object might also be returned if there is a
+            # single geometry field
+            return {'my_fid': self.iter,
+                    'boolField': True,
+                    'intField': self.iter + 1,
+                    'int64Field': 1234567890123,
+                    'realField': 1.23,
+                    'strField': 'foo',
+                    'strNullField': None,
+                    'binaryField': b'\x01\x00\x02',
+                    'datetimeField': '2017-04-26T12:34:56.789Z',
+                    'geomField': 'POINT(2 49)'}
+
+
+    class Dataset(GDALDataset):
+
+        def __init__(self, filename):
+            self.layers = [Layer()]
+            pass
+
+        # Optional, called on native object destruction
+        def __del__(self):
+            pass
+
+        # Optional
+        def metadata(self, domain):
+            if domain is None:
+                return {'foo': 'bar'}
+            return None
+
+        # Required, unless a layers attribute is set in __init__
+        # def layer_count(self):
+        #     return len(self.layers)
+
+        # Required, unless a layers attribute is set in __init__
+        # def layer(self, idx):
+        #     return self.layers[idx]
+
+
+    class Driver(GDALDriver):
+
+        # Required
+        def identify(self, filename, first_bytes, open_flags):
+            return filename == 'MYPLUGIN:'
+
+        # Required
+        def open(self, filename, first_bytes, open_flags):
+            if not self.identify(filename, first_bytes, open_flags):
+                return None
+            return Dataset(filename)
+
+
+    # Required
+    def create_driver():
+        return Driver()
```

Comment:
Why this `create_driver` is required; what's the purpose?
Why not to use the `__init__` magic?

Since user's module is regularly imported
 https://github.com/rouault/gdal/blob/1bf795f25458dfb7e2a88e5d99b0e3647b82f2d4/gdal/gcore/gdalpythondriverloader.cpp#L2212-L2214

then why not look for classes that are `PyType_IsSubtype` of `Driver` class and construct calling the regular Python `__init__`?

```cpp
PyObject* poTypeDriver = ... ; // class item found with PyDict_Next for callable PyType_Check
                               // that is PyType_IsSubtype of Driver
PyTypeObject* poDriver = reinterpret_cast<PyTypeObject*>(poTypeDriver);
if (poDriver->tp_new(obj.get(), args, kwds) < 0) { ... }
if (poDriver->tp_init(obj.get(), args, kwds) < 0) { ... }

PyObject* poMethod = PyObject_GetAttrString(poDriver, "open" );
...

```

Reply:
sounds like a good idea. Adding a bit of implementation complexity though. Do you have an idea how to enumerate the classes of a module ? Can't see anything better than PyModule_GetDict() ( https://docs.python.org/3/c-api/module.html#c.PyModule_GetDict ) although they suggest using other methods, but not clear to me which one(s) would be appropriate

Reply:
I usually do something along this lines (read the snippet as pseudo-code, not compile/run-tested, written from brain memory):

```cpp
PyObject *pKey(nullptr), *pValue(nullptr);
Py_ssize_t pos(0); // must be initialized to 0 prior call to PyDict_Next! 

PyObject* pModule = ...; // module with Driver, OGRLayer, etc.
PyObject* pDriverClass = PyObject_GetAttrString(pModule, "Driver");
PyTypeObject* pDriverType = reinterpret_cast<PyTypeObject*>(pDriverClass);

PyObject* pDict = PyModule_GetDict(pPluginModule, ...);
while (PyDict_Next(pDict, &pos, &pKey, &pValue))
{
    PyObject* pDictAttr = PyObject_GetAttr(pDict, pKey)
    if (PyType_Check(pDictAttr))
    {
        PyTypeObject* pType = reinterpret_cast<PyTypeObject*>(pDictAttr);
        if (PyType_IsSubtype(pDriverType, pType))
        {
            // User-defined Python class is viable candidate for Driver class
            // ... allocate/instantiate/call it...
        }
    }
}
```

> they suggest using other methods, but not clear to me which one(s) would be appropriate

I think, for `PyObject_*` they mean functions like `PyObject_GetAttrString`, but for `PyModule_*`, not sure.

Reply:
ok, I've managed to no longer require a create_driver() method. Using Python C API wasn't really possible as most of the PyType_xxxx methods are actually macros, which require having knowledge of Python structures, which isn't super ideal given that I don't include Python.h to be as much version independent as possible... So I went through the road of executing a Python snippet instead: https://github.com/rouault/gdal/commit/884ced57d8e720793e9f65f9eb059276d5db1c32#diff-42d4e6b0795db7beb7d99c0b9e51b18fR1190

Reply:
>  PyType_xxxx methods are actually macros, which require having knowledge of Python structures (...) I don't include Python.h ...

Sorry, I ignored that fact indeed.